localization_manager/src/LocalizationTransitionTable.cpp

```diff
@@ -0,0 +1,230 @@
+                                                                                 << state_);
+}
+
+void LocalizationTransitionTable::setAndLogState(LocalizationState new_state, LocalizationSignal source_signal)
+{
+  if (new_state == state_)
+  {
+    return;  // State was unchanged no need to log or trigger callbacks
+  }
+  ROS_INFO_STREAM("LocalizationTransitionTable changed localization state from "
+                  << state_ << " to " << new_state << " because of signal " << source_signal << " while in mode "
+                  << mode_);
+  LocalizationState prev_state = state_;
+  state_ = new_state;
+  if (transition_callback_)
+  {
+    transition_callback_(prev_state, state_, source_signal);
+  }
+}
+
+void LocalizationTransitionTable::signalWhenUNINITIALIZED(LocalizationSignal signal)
+{
+  switch (signal)
+  {
+    case LocalizationSignal::INITIAL_POSE:
+      if (mode_ == LocalizerMode::GNSS)
+      {
+        setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
+      }
+      else
+      {
+        setAndLogState(LocalizationState::INITIALIZING, signal);
+      }
+      break;
+    default:
+      logDebugSignal(signal);
+      break;
+  }
+}
+
+void LocalizationTransitionTable::signalWhenINITIALIZING(LocalizationSignal signal)
+{
+  switch (signal)
+  {
+    // How to handle the combined conditions?
+    case LocalizationSignal::GOOD_NDT_FREQ_AND_FITNESS_SCORE:
+      setAndLogState(LocalizationState::OPERATIONAL, signal);
+      break;
+    case LocalizationSignal::POOR_NDT_FREQ_OR_FITNESS_SCORE:
+      setAndLogState(LocalizationState::DEGRADED, signal);
+      break;
+    case LocalizationSignal::LIDAR_SENSOR_FAILURE:
+      setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
+      break;
+    case LocalizationSignal::TIMEOUT:
+      setAndLogState(LocalizationState::AWAIT_MANUAL_INITIALIZATION, signal);
+      break;
+    default:
+      logDebugSignal(signal);
+      break;
+  }
+}
+void LocalizationTransitionTable::signalWhenOPERATIONAL(LocalizationSignal signal)
+{
+  switch (signal)
+  {
+    case LocalizationSignal::INITIAL_POSE:
+      setAndLogState(LocalizationState::INITIALIZING, signal);
+      break;
+    case LocalizationSignal::POOR_NDT_FREQ_OR_FITNESS_SCORE:
+      setAndLogState(LocalizationState::DEGRADED, signal);
+      break;
+    case LocalizationSignal::LIDAR_SENSOR_FAILURE:
+      if (mode_ == LocalizerMode::AUTO)
+      {
+        setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
+      }
+      else
+      {
+        setAndLogState(LocalizationState::AWAIT_MANUAL_INITIALIZATION, signal);
+      }
+      break;
+    case LocalizationSignal::UNUSABLE_NDT_FREQ_OR_FITNESS_SCORE:
+      if (mode_ == LocalizerMode::AUTO)
+      {
+        setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
+      }
+      else
+      {
+        setAndLogState(LocalizationState::AWAIT_MANUAL_INITIALIZATION, signal);
```

Comment:
LocalizationState::AWAIT_MANUAL_INITIALIZATION,
LocalizationSignal::LIDAR_SENSOR_FAILURE
seem to have same operation, should we combine them to make fewer lines for testing, or is this way more clear?

Reply:
I'll make it a fall through case